subsys/nrf_compress/src/lzma.c

```diff
@@ -303,6 +422,260 @@ static int lzma_decompress(void *inst, const uint8_t *input, size_t input_size,
+	decoder = &lzma_decoder.decoder;
+#else
+	decoder = &lzma_decoder;
+#endif
+
+	if (!allocated_probs) {
+#ifdef CONFIG_NRF_COMPRESS_LZMA_VERSION_LZMA2
+		rc = Lzma2Dec_AllocateProbs(&lzma_decoder, input[0], &lzma_probs_allocator);
+#else
+		rc = LzmaDec_AllocateProbs(&lzma_decoder, input, LZMA_PROPS_SIZE,
+					   &lzma_probs_allocator);
+#endif
+
+		if (rc) {
+			rc = -EINVAL;
+			goto done;
+		}
+
+		decoder->dicHandle = LzmaDictionaryOpen(decoder->prop.dicSize);
+
+		if (decoder->dicHandle == NULL) {
+			rc = -EINVAL;
+#ifdef CONFIG_NRF_COMPRESS_LZMA_VERSION_LZMA2
+			Lzma2Dec_FreeProbs(&lzma_decoder, &lzma_probs_allocator);
+#else
+			LzmaDec_FreeProbs(&lzma_decoder, &lzma_probs_allocator);
+#endif
+			goto done;
+		}
+
+		allocated_probs = true;
+#ifdef CONFIG_NRF_COMPRESS_LZMA_VERSION_LZMA2
+		*offset = LZMA2_HEADER_SIZE;
+#else
+		/* Header and account for uncompressed size */
+		*offset = LZMA_PROPS_SIZE + sizeof(uint64_t);
+#endif
+
+#ifdef CONFIG_NRF_COMPRESS_LZMA_VERSION_LZMA2
+		Lzma2Dec_Init(&lzma_decoder);
+#else
+		LzmaDec_Init(&lzma_decoder);
+#endif
+
+		return 0;
+	}
+
+#ifdef CONFIG_NRF_COMPRESS_LZMA_VERSION_LZMA2
+	rc = Lzma2Dec_DecodeToDic(&lzma_decoder, MAX_LZMA_DICT_SIZE, input, &chunk_size,
+					(last_part ? LZMA_FINISH_END : LZMA_FINISH_ANY), &status);
+#else
+	rc = LzmaDec_DecodeToDic(&lzma_decoder, MAX_LZMA_DICT_SIZE, input, &chunk_size,
+					(last_part ? LZMA_FINISH_END : LZMA_FINISH_ANY), &status);
+#endif
+	if (rc) {
+		rc = -EINVAL;
+		goto done;
+	}
+
+	*offset = chunk_size;
+
+	if (last_part && (status == LZMA_STATUS_FINISHED_WITH_MARK ||
+			  status == LZMA_STATUS_MAYBE_FINISHED_WITHOUT_MARK) &&
+	    *offset < input_size) {
+		/* If last block, ensure offset matches complete file size */
+		*offset = input_size;
+	}
+
+	if (decoder->dicPos >= decoder->dicHandle->dicBufSize ||
+	    (last_part && input_size == chunk_size)) {
+#if CONFIG_NRF_COMPRESS_DICTIONARY_CACHE_SIZE > 0
+		if (cache.invalid) {
+			validate_cache(decoder->dicHandle);
+		}
+#endif
+		*output_size = decoder->dicPos;
+		decoder->dicPos = 0;
+	}
+
+done:
+	return rc;
+}
+
+static DictHandle dict_handle;
+
+DictHandle *LzmaDictionaryOpen(SizeT size)
+{
+	if (dict_handle.isOpened)
+		return &dict_handle;
+
+	if (ext_dict ==  NULL)
+		return NULL;
+
+	size_t dict_size;
+
+	if (ext_dict->open((size_t)size, &dict_size) != 0) {
+		LOG_WRN("Unable to open external dictionary with size %u", size);
+		return NULL;
+	}
+
+	dict_handle.isOpened = True;
+	dict_handle.dicBufSize = dict_size;
+
+#if CONFIG_NRF_COMPRESS_DICTIONARY_CACHE_SIZE > 0
+	cache.dict_pos_begin = 0;
+	cache.dict_pos_end = sizeof(cache.data) - 1;
+#endif
+
+	return &dict_handle;
+}
+
+SizeT LzmaDictionaryWrite(DictHandle *handle, SizeT pos, const Byte *data, SizeT len)
+{
+	if (handle != &dict_handle || ext_dict == NULL)
+		return 0;
+
+	SizeT write_len = len;
+
+	if (pos + len > handle->dicBufSize) {
+		write_len = handle->dicBufSize - pos;
```

Comment:
1) Not sure if possible to happen, but for safety reason best add a check that `handle->dicBufSize >= pos`

2) In case `CONFIG_NRF_COMPRESS_DICTIONARY_CACHE_SIZE == 0` and we do not enter the part below - how is wrapping of the dictionary performed? Won't we stay in `pos==handle->dicBufSize` forever?